Broadcast and filter game options

11EJDE11 opened the pull request:
Closes or helps close

- #816 
- https://github.com/CnCNet/cncnet-yr-client-package/issues/611 
- https://github.com/CnCNet/cncnet-yr-client-package/issues/278
  
--------

- Adds game options to the GAME broadcast
- Adds broadcasted game options to the filters in CnCNetLobby
- Adds broadcasted game options to the GameListBox and GameInformationPanel
- Adds setting to disable the game icon from showing in the GameListBox

**New INI settings for GameSessionCheckBox (located in `GameLobbyBase.ini`):**
```ini
BroadcastToLobby=Yes|No
ShowInGameList=Yes|No
ShowInGameListOnRight=Yes|No
ShowInGameInformationPanel=Yes|No
ShowInGameInformationPanelAsIconOnly=Yes|No
ShowIconInGameLobby=Yes|No
ShowInFilters=Yes|No
EnabledIcon=<filename.png>
DisabledIcon=<filename.png>
SortOrder=<integer>
```
New settings for GameSessionDropDown (located in GameLobbyBase.ini):
```ini
BroadcastToLobby=Yes|No
ShowInGameList=Yes|No
ShowInGameListOnRight=Yes|No
ShowInGameInformationPanel=Yes|No
ShowInGameInformationPanelAsIconOnly=Yes|No
ShowIconInGameLobby=Yes|No
ShowInFilters=Yes|No
Icons=<icon1.png>,<icon2.png>,...
SortOrder=<integer>
```

**New setting in `ClientDefinitions.ini`:**
```ini
[Settings]
ShowGameIconInGameList=Yes|No
```

**GameListBox and GameInformationPanel changes:**
<img width="834" height="616" alt="image" src="https://github.com/user-attachments/assets/6a2e5c8d-40c1-46f8-a22c-a66fcdae1836" />


**Filters menu:**
<img width="536" height="637" alt="image" src="https://github.com/user-attachments/assets/9f4e3a5e-c394-4324-b78f-972b3b81f7c3" />

11EJDE11 committed:
Add icon legend to GameInformationPanel
Fix issue with lblSkillLevel visibility
Add layout constants

Co-authored-by: devo1929 <devo1929@users.noreply.github.com>

11EJDE11 committed:
Update using statements for alphabetical order
Add extra param to CreateDivider to specify a custom height
Add constant for spacing in icon panel

11EJDE11 committed:
Add game options to the GAME broadcast
Add game options to the filters in CnCNetLobby
Add broadcasted game options to the GameListBox and GameInformationPanel
Adds setting to disable the game icon from the GameListBox

Metadorius commented on DXMainClient/DXGUI/Multiplayer/CnCNet/CnCNetLobby.cs:
```diff
@@ -1551,6 +1616,43 @@ private void GameBroadcastChannel_CTCPReceived(object sender, ChannelCTCPEventAr
+                            int bitIndex = i % 32;
+                            if (packedIndex < packedCheckboxes.Length)
+                                checkboxValues[i] = (packedCheckboxes[packedIndex] & (1 << bitIndex)) != 0;
+                        }
```
could you explain what is happening here? I see bitpacking which is a clever optimisation, but apart from that I have no idea

Metadorius commented on DXMainClient/DXGUI/Multiplayer/GameFiltersPanel.cs:
```diff
@@ -153,7 +382,29 @@ private void Save()
+                }
+                else
+                {
+                    // For dropdowns: -1 = All (default), actual indices are offset by 1
```
Not sure what is the reason for that math here?

11EJDE11 replied:
Sorted with [60b23c5](https://github.com/CnCNet/xna-cncnet-client/pull/874/commits/60b23c56f4fe4fab0d3dd81ecc501b26701141ce).

Metadorius commented on DXMainClient/DXGUI/Generic/GameSessionCheckBox.cs:
```diff
@@ -57,7 +57,43 @@ public bool AllowScoring
+    /// <summary>
+    /// Whether the icon should be shown in the game filters panel.
+    /// </summary>
+    public bool IconShownInFilters { get; private set; }
```
Hmm, I know I said previously that it would be good to have it customisable, but I am not sure now. Do you think there would be any case when the modder or game maintainer would want to have an icon in game list but *not* have the icon anywhere else? ATM I am not sure if this makes sense, because the game info/filters icon work as legend, so probably could omit that.

Also I think we could draw the icons near game options in the lobby themselves the same way?

11EJDE11 replied:
Yeah, I think we keep the settings but make it apply to the whole thing not just the icon. So "ShownInGameInfo" instead of "IconShownInGameInfo", etc. It's programmed a bit like that already - hence the missing Credits in that screenshot.

I was having hassles with icons in the game lobby. I'll take another look.

Metadorius replied:
I think the filter and the game info should be switched on by the same option, otherwise it would be pretty user unfriendly to have something in filter but not in game info, and vice versa. Same for icons in the list - if the icon is shown in the game list - it must have the same icon (working as legend) in filter and in game info, so the users don't get confused on the meaning.

Metadorius reviewed (changes requested):
Good work so far mate!

<img width="152" height="231" alt="image" src="https://github.com/user-attachments/assets/dd4487cd-5c85-40e3-b61a-19d10bb0bcbd" />

<img width="167" height="88" alt="image" src="https://github.com/user-attachments/assets/1db80315-8164-4750-8f66-7604001c816d" />

I don't see if credits are shown in the game info panel, is this intended? I thought if you're able to filter by it - it should also be visible there.

Also maybe the panel should be made two-column and use the same font as the main game info?

Also I think the icons here should be *item* icons, because so far per the idea icons convey the *value*, not *what option* it is. See my review comments on how to make this work.

<img width="158" height="65" alt="image" src="https://github.com/user-attachments/assets/98c69166-bae5-4285-9227-820f07fea63e" />

Some of the comments that I left are more of open questions, so curious to hear your thoughts on them.

11EJDE11 committed:
Change saved filters to nullable

Metadorius commented on DXMainClient/DXGUI/Generic/GameSessionCheckBox.cs:
```diff
@@ -57,7 +57,43 @@ public bool AllowScoring
+    /// Whether the icon should be shown on the right side of the game list.
+    /// Only applies if IconShownInGameList is true.
+    /// </summary>
+    public bool IconShownInGameListOnRight { get; private set; }
```
Not sure, is there a use for having both?

11EJDE11 committed:
Refactor game options to use unified integer-based interface

11EJDE11 commented on DXMainClient/DXGUI/Multiplayer/CnCNet/CnCNetGameLoadingLobby.cs:
```diff
@@ -717,8 +717,9 @@ private void BroadcastGame()
-            sb.Append(string.Empty); // MapHash
+            sb.Append(";"); // Map SHA1
+            sb.Append(";"); // Packed checkbox values (empty for loaded lobbies)
+            sb.Append(";"); // Dropdown indices (empty for loaded lobbies)
```
Not needed. That stuff was mostly copied from the existing game options broadcast which had it split. See: https://github.com/CnCNet/xna-cncnet-client/blob/3cf078ba89fd59eb51ff661a7cf246f3696fa995/DXMainClient/DXGUI/Multiplayer/GameLobby/CnCNetGameLobby.cs#L1036

RE the interface - makes sense. I was working on an outdated copy and didn't see those changes. Hopefully [60b23c5](https://github.com/CnCNet/xna-cncnet-client/pull/874/commits/60b23c56f4fe4fab0d3dd81ecc501b26701141ce) sorts it as you imagined.

11EJDE11 commented on DXMainClient/DXGUI/Multiplayer/CnCNet/CnCNetLobby.cs:
```diff
@@ -460,6 +469,62 @@ private bool HostedGameMatches(GenericHostedGame hg)
+                    if (gameSelectedIndex != filterValue)
+                        return false;
+                }
+            }
```
[60b23c5](https://github.com/CnCNet/xna-cncnet-client/pull/874/commits/60b23c56f4fe4fab0d3dd81ecc501b26701141ce) sorts it.

11EJDE11 committed:
Fix duplicate L10N

11EJDE11 committed:
Switch to regular font instead of bold

11EJDE11 committed:
Add draw items in row on GameInformationPanel
Add 2 column drawing in GameInformationPanel
Add SortOrder property
Add icons to dropdowns
Rename settings
Use scrollpanel

11EJDE11 committed:
Fix icons in dropdowns in filter panel

Metadorius commented on DXMainClient/DXGUI/Generic/GameSessionDropDown.cs:
```diff
@@ -39,19 +39,48 @@ public GameSessionDropDown(WindowManager windowManager) : base(windowManager) {
+    /// Sort order for displaying icons in the GameInformationPanel and GameListBox.
+    /// Lower values appear first.
+    /// </summary>
+    public int SortOrder { get; private set; } = 100;
```
Not sure why this is 100 😅

SadPencil replied:
So people can easily specify a lower or higher control? Seems like a common way, like the default metric for a network interface.

Mabybe we can make this `100` a constant but `100` is definitely better than a default 0 (unless we want to allow negative numbers, then `0` will be a perfect default value) or `int.Max`

Metadorius replied:
I don't see why would we disallow negative values

11EJDE11 replied:
I've changed it to 0 as I'm not blocking negatives so I guess it makes more sense. Setting it to 100 just read better to me.

SadPencil commented on DXMainClient/DXGUI/Multiplayer/GameInformationPanel.cs:
```diff
@@ -221,6 +261,213 @@ public void SetInfo(GenericHostedGame game)
+            // First show icons in a row
+            if (optionIconsOnly.Count > 0)
+            {
+                var sortedIconsOnly = optionIconsOnly.OrderBy(x => x.sortOrder).ToList();
```
Nitpick. The `List<T>.Sort` method sorts the list in-place and has a higher efficiency than an `.OrderBy.ToList()`.

(It's a pity that PriorityQueue is only provided in .NET 6+ but sorting a list here is totally acceptable)

11EJDE11 replied:
Thanks SP. Let's leave it as is for now. It's not sorting much and would require a bit more code otherwise.

SadPencil commented on DXMainClient/DXGUI/Multiplayer/GameInformationPanel.cs:
```diff
@@ -221,6 +261,213 @@ public void SetInfo(GenericHostedGame game)
+                currentY += maxIconHeight + legendPadding;
+            }
+
+            // Then show icons with text ( two columns)
```
`( two columns)` -> `(two columns)`

11EJDE11 replied:
Fixed thanks.

SadPencil commented on DXMainClient/DXGUI/Multiplayer/GameInformationPanel.cs:
```diff
@@ -221,6 +261,213 @@ public void SetInfo(GenericHostedGame game)
+                    pnlGameOptions.AddChild(iconPanel);
+
+                    if (isLeftColumn)
+                        leftY += legendIconHeight + 2;
```
You have defined `const int iconSpacing = 4; private const int legendPadding = 5;` but here we got another magic number `2`. You might want to define another constant

11EJDE11 replied:
Thanks, have added a constant.

SadPencil commented on ClientCore/Settings/UserINISettings.cs:
```diff
@@ -433,6 +469,26 @@ public void ResetGameFilters()
+        {
+            var section = SettingsIni.GetSection(GAME_OPTION_FILTERS);
+            if (section != null)
+                section.RemoveAllKeys();
```
section?.RemoveAllKeys();

11EJDE11 replied:
Fixed, thanks.

SadPencil commented on DXMainClient/DXGUI/Multiplayer/GameLobby/CnCNetGameLobby.cs:
```diff
@@ -2087,6 +2087,39 @@ private void BroadcastGame()
+
+                // Convert bytes to integers
+                for (int i = 0; i < byteArray.Length / 4; i++)
+                    gameOptionValues.Add(BitConverter.ToInt32(byteArray, i * 4));
```
Same as the V3 PR: be careful using BitConverter. You can throw if `BitConverter.IsLittleEndian` does not match our expect, or use `BinaryPrimitives` instead

SadPencil replied:
Update: looks like we already have ~20 `BitConverter` usages. Maybe we should change them to `BinaryPrimitives` or globally throw on unexpected `BitConverter.IsLittleEndian` in another PR

11EJDE11 replied:
OK - let's do this in another PR.

SadPencil commented on DXMainClient/DXGUI/Multiplayer/GameInformationIconPanel.cs:
```diff
@@ -0,0 +1,54 @@
+            int textHeight = (int)textSize.Y;
+
+            int iconY = (textHeight - icon.Height) / 2;
+            if (iconY < 0) iconY = 0;
```
Our style requires a line break after `if (condition)`

11EJDE11 replied:
Sorted, thanks.

SadPencil reviewed (commented):
LGTM. Only some minor comments

11EJDE11 committed:
Merge remote-tracking branch 'origin/develop' into feat-broadcast-game-settings

11EJDE11 committed:
Fix formatting

11EJDE11 committed:
Add SortOrder constant

11EJDE11 committed:
Fix null check

11EJDE11 committed:
Add constant for padding

11EJDE11 committed:
Fix spacing

11EJDE11 committed:
Make default sort order 0

11EJDE11 committed:
Restore other games icons

11EJDE11 commented on DXMainClient/DXGUI/Generic/GameSessionDropDown.cs:
```diff
@@ -33,6 +33,26 @@ public GameSessionDropDown(WindowManager windowManager) : base(windowManager) {
+    /// <summary>
+    /// The texture name for the icon.
+    /// </summary>
+    public string Icon { get; private set; }
```
Thanks and fixed.

11EJDE11 commented on DXMainClient/DXGUI/Multiplayer/GameInformationPanel.cs:
```diff
@@ -221,6 +261,161 @@ public void SetInfo(GenericHostedGame game)
+                            Texture2D icon = AssetLoader.LoadTexture(iconName);
+                            if (icon != null)
+                            {
+                                string text = $"{checkbox.Text}: {(isChecked ? "On".L10N("Client:Main:On") : "Off".L10N("Client:Main:Off"))}";
```
I think on/off is more appropriate as we're displaying the status of an option not answering a yes/no question. We do use yes/no throughout the ini files though.
Happy to change it if you want, I have no strong preference either way.

-------------

SW: On
Crates: Off

-------------

SW: Yes
Crates: No

-------------

11EJDE11 commented on DXMainClient/DXGUI/Multiplayer/GameFiltersPanel.cs:
Thanks, it is using the scroll panel now. There is an annoying issue where scrolling with the mousewheel, if the cursor is over a dropdown, the dropdown scrolls too.
